### PR TITLE
Add GPTOSS Example with Megatron-LM + Megatron Bridge

### PR DESCRIPTION
**Pull request by faradawn**
# What does this PR do ?

This PR adds examples for
- Convert checkpoints from HF to Megatron format using Megatron Bridge
- Train GPT-OSS with Megatron-LM
- Convert the final Megatron checkpoint to HF format.

Tested on B300 with 8 GPUS.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by faradawn** [comment]
Hi @sbhavani this PR is ready. Do you know how to added the "Expert Review" Label? I cannot find a "gear icon". cc @AlphaBladez

File: examples/gptoss/README.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+```bash
+export HF_TOKEN={your_hf_token_here}
+```
+
+## Step 1: Convert HuggingFace to Megatron (Optional - skip if you already have a Megatron checkpoint)
+
+Set `--nproc-per-node` to be the number of GPUs per node. Set `hf_model_name` to be the Huggingface model e.g. `openai/gpt-oss-20b`
+
+```bash
+python3 -m torch.distributed.launch --nproc-per-node=8 examples/gptoss/01_convert_from_hf.py --hf-model openai/gpt-oss-20b
```
Comment (sbhavani) [inline examples/gptoss/README.md:62]: instead of torch.distributed.launch use torchrun

File: examples/gptoss/README.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+# GPT-OSS Training Tutorial
+
+## Step 0: Install Dependencies
+
+### Using Megatron Bridge
+
+[Megatron-Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge)
+
+Megatron bridge provides a quick and convenient way to convert HuggingFace checkpoints to the Megatron format used by Megatron-LM. Follow the instructions in the [Megatron-Bridge Installation](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/README.md#-installation) to run the nemo docker container and convert checkpoints (via mounted volumes - make sure that the huggingface cache location AND the megatron checkpoint locations are properly mounted, otherwise you may not be saving the converted model to disk correctly).
```
Comment (sbhavani) [inline examples/gptoss/README.md]: Consistency: "Megatron Bridge"

File: examples/gptoss/README.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+Set `--nproc-per-node` to be the number of GPUs per node. Set `hf_model_name` to be the Huggingface model e.g. `openai/gpt-oss-20b`
+
+```bash
+python3 -m torch.distributed.launch --nproc-per-node=8 examples/gptoss/01_convert_from_hf.py --hf-model openai/gpt-oss-20b
+```
+
+## Step 2: Train from Scratch
+
+To train from scratch first follow the steps below to setup the environment appropriately before running the training script in docker. Even though we are running the same container as before, it is better to restart the container to ensure a clean environment and that all environment and docker variables are set correctly.
```
Comment (sbhavani) [inline examples/gptoss/README.md]: Mention GPU used here and number.

File: examples/gptoss/02_train.sh [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    --nnodes $NUM_NODES
+    --master_addr $MASTER_ADDR
+    --master_port $MASTER_PORT
+    --node_rank $NODE_RANK
+)
+
+# NOTE: Adjust the following model and training parameters as needed below, these are example values for openai/gpt-oss-20b
+TP_SIZE=1     
+EP_SIZE=1     
+PP_SIZE=1
```
Comment (sbhavani) [inline examples/gptoss/02_train.sh]: Conversion script sets PP_SIZE=WORLD_SIZE but training script defaults to PP_SIZE=1

File: examples/gptoss/02_train.sh [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    --node_rank $NODE_RANK
+)
+
+# NOTE: Adjust the following model and training parameters as needed below, these are example values for openai/gpt-oss-20b
+TP_SIZE=1     
+EP_SIZE=1     
+PP_SIZE=1     
+MICRO_BATCH_SIZE=1
+GLOBAL_BATCH_SIZE=128
+NUM_LAYERS=12
```
Comment (sbhavani) [inline examples/gptoss/02_train.sh:118]: Should mention in README.md that this is modified for testing purposes and the actual model has more layers. Same with hidden size

File: examples/gptoss/README.md (hunk 1 above)
Comment (AlphaBladez) [reply examples/gptoss/README.md:62]: `torchrun` is the preferred way of doing this, but unfortunately does not play nicely with `uv` and cannot pickup the environment we created in the steps before. We will note this in a comment, but leave in `torch.distributed.launch` for now.

File: examples/gptoss/02_train.sh (hunk 5 above)
Comment (faradawn) [reply examples/gptoss/02_train.sh]: Notes added to README

File: examples/gptoss/02_train.sh (hunk 4 above)
Comment (faradawn) [reply examples/gptoss/02_train.sh]: I saw that Sachin has adjusted training script to PP_SIZE = world size

File: examples/gptoss/README.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+NUM_NODES=1
+MASTER_ADDR=localhost
+MASTER_PORT=6000
+NODE_RANK=0
+EOF
+```
+
+### Run Container with Mounted Volumes
+
+**NOTE:** This container runs the example training script `02_train.sh` located in the `examples/gptoss` directory. By default, we have only set pipeline parallelism to be the number of GPUs. Adjust TP_SIZE, EP_SIZE, PP_SIZE, etc. in `02_train.sh`. You can also adjust modify `--hidden-size`, `--ffn-hidden-size`, `--num-attention-heads`, `NUM_LAYERS`, etc.
```
Comment (faradawn) [inline examples/gptoss/README.md]: @sbhavani can you check if this note in README looks good?
